### ISSUE TITLE
Update to Mercurial 3.5 breaks hg push of Tool Shed repositories

### ISSUE THREAD

**Issue by nsoranzo** [galaxyproject/galaxy#1183]
In Mercurial 3.5 (the wheel required after #428) the bundle format changed [from 1.0 to 2.0](https://www.mercurial-scm.org/wiki/BundleFormat2) . This breaks the `tool_shed.util.hg_util.bundle_to_json()` function, which is needed by the `Hg` middleware defined in `lib/galaxy/webapps/tool_shed/framework/middleware/hg.py` .

This can be easily verified by running:

```
./run_tests.sh -toolshed test/tool_shed/functional/test_0310_hg_push_from_api.py
```

The following patch fixes the first error, but a major rework would be needed to fix the other problems.

```
diff --git a/lib/tool_shed/util/hg_util.py b/lib/tool_shed/util/hg_util.py
index ac01ca7..be395a3 100644
--- a/lib/tool_shed/util/hg_util.py
+++ b/lib/tool_shed/util/hg_util.py
@@ -44,7 +44,7 @@ def bundle_to_json( fh ):
     command line) to a json object.
     """
     # See http://www.wstein.org/home/wstein/www/home/was/patches/hg_json
-    hg_unbundle10_obj = readbundle( None, fh, None )
+    hg_unbundle10_obj = readbundle( get_configured_ui(), fh, None )
     groups = [ group for group in unpack_groups( hg_unbundle10_obj ) ]
     return json.dumps( groups, indent=4 )

```

**Comment by martenson** [comment]
I think the best solution in this case is fixing on version 3.2.4. I doubt that we gain enough from switching 3.2 > 3.5 to justify adaptation.

**Comment by nsoranzo** [comment]
The last working Mercurial version is 3.4.2 . The best solution to me is to pin this version and deprecate `hg push` of Tool Shed repositories.

**Comment by martenson** [comment]
@nsoranzo this is still in fact upgrading mercurial from 3.2.4 to 3.4.2, right?

**Comment by nsoranzo** [comment]
@martenson Yes, if it doesn't work for some reason we can always downgrade more.

**Comment by martenson** [comment]
alright, I am :+1: on this

**Comment by nsoranzo** [comment]
`hg push` have been deprecated in #1194.

@martenson When do you think it will be appropriate to remove the `hg push` code?

**Comment by martenson** [comment]
@nsoranzo I am fine next release; whether we find the manpower to do it is unclear though

**Comment by natefoo** [comment]
After some Starforge bug fixes and updates to the images, 3.4.2 wheels are up on wheels.galaxyproject.org. All that should be necessary now is to modify the version in requirements.txt.

**Comment by nsoranzo** [comment]
Thanks @natefoo, I will submit a PR later today.
